File: rss/src/rss.rs
use rss::{CategoryBuilder, ChannelBuilder, GuidBuilder, Item, ItemBuilder};

pub trait RssItem {
    fn get_title(&self) -> String;
    fn get_link(&self) -> String;
    fn get_description(&self) -> String;
    fn get_guid(&self) -> String;
    fn get_author(&self) -> String;
    fn get_category(&self) -> String;
    fn get_pubdate(&self) -> String;
}

pub struct RssChannelConfig<'a> {
    pub title: &'a str,
    pub link: &'a str,
    pub description: &'a str,
    pub language: Option<&'a str>,
}

pub fn generate_rss<T: RssItem>(config: &RssChannelConfig, items: &Vec<T>) -> String {
    let mut rss_items: Vec<Item> = Vec::new();

    for item in items {
        let guid = GuidBuilder::default()
            .value(item.get_guid())
            .permalink(false)
            .build();

        let category = CategoryBuilder::default().name(item.get_category()).build();

        let rss_item = ItemBuilder::default()
            .title(item.get_title())
            .link(item.get_link())
            .description(item.get_description())
            .author(item.get_author())
            .guid(Some(guid))
            .categories(vec![category])
            .pub_date(item.get_pubdate())
            .build();

        rss_items.push(rss_item);
    }

    ChannelBuilder::default()
        .title(config.title)
        .link(config.link)
        .description(config.description)
        .items(rss_items)
        .build()
        .to_string()
}

File: src/rss.ts
export interface RssChannelConfig {
  title: string;
  link: string;
  language: string | null;
}

export interface RssItem {
  title: string;
  link: string;
  author: string;
  pub_date: string;
  guid: { value: string; is_permalink: boolean };
  description: string;
}

export function buildRss(config: RssChannelConfig, items: RssItem[]): string {
  const rssitem = items.map((item): string => {
    return (
      "<item>" +
      `<title>${item.title}</title>` +
      `<link>${item.link}</link>` +
      `<author>${item.author}</author>` +
      `<pubDate>${item.pub_date}</pubDate>` +
      `<guid isPermaLink=${item.guid.is_permalink}>${item.guid.value}</guid>` +
      `<description><![CDATA[${item.description}]]></description></item>`
    );
  });

  return (
    `<rss version="2.0">` +
    "<channel>" +
    `<title>${config.title}</title>` +
    `<link>${config.link}</link>` +
    `<language>${config.language ?? ""}</language>` +
    rssitem.join("") +
    "</channel></rss>"
  );
}
